### PR TITLE
Added UART bridge test

### PR DESCRIPTION
**Pull request by nandor**
Added a UART test in a new group of tests, 'bridges'. This test group includes the bridges from firesim-lib, which rely on more external dependencies. The previous `midasexamples` test suite is limited to simple tests with no dependencies.
The `PointerChaser` test will have to be moved to the `fasedtests` group later on.

### Contributor Checklist
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [x] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you state the UI / API impact?
- [x] Did you specify the Verilog / AGFI compatibility impact?

- [ ] If applicable, did you regenerate and publicly share default AGFIs?

- [ ] If applicable, did you apply the `ci:fpga-deploy` label?

- [ ] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
Note: to run CI on PRs from forks, comment `@Mergifyio copy main` and manage the change from the new PR.
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

**Comment by davidbiancolin** [comment]
FWIW, i've been meaning to remove the pointer chaser. It's not a good enough test to warrant maintain the various oddities it has.

File: .github/workflows/firesim-run-tests.yml
Comment (davidbiancolin) [inline .github/workflows/firesim-run-tests.yml:259]: For now refer to these as `firesim-lib` bridges. We started talking about renaming these to berkeley-bridges or something more informative, but i need to pull all ofthe non-bridge stuff out of firesim-lib first.

File: .github/workflows/firesim-run-tests.yml
Comment (davidbiancolin) [inline .github/workflows/firesim-run-tests.yml:267]: Update this

File: sim/firesim-lib/src/main/cc/bridges/uart.cc
Comment (davidbiancolin) [inline sim/firesim-lib/src/main/cc/bridges/uart.cc:92]: please add `ci:fpga-deploy" just to sanity check things continue to work correctly.

File: sim/firesim-lib/src/main/cc/bridges/uart.h
Comment (davidbiancolin) [inline sim/firesim-lib/src/main/cc/bridges/uart.h:22]:
```suggestion
 * Base class for callbacks handling data coming in and out of our UART stream.
```
This is what you meant right?

File: sim/src/main/cc/bridges/UARTModule.cc
Comment (davidbiancolin) [inline sim/src/main/cc/bridges/UARTModule.cc:57]: How long does this test take to run? For the string you're writing you could make this test run much more quickly by calling tick more frequently relative to step. For testing bridges like this, you can favor the non-blocking variant of "step" and instead write:

```
step(<some_reasonable_upper_bound>, false)
while(!done() <(&& possibly other conditions)>) {
  uart->tick();
}
```

Comment (davidbiancolin) [reply sim/src/main/cc/bridges/UARTModule.cc:57]: Basically the UART is sitting idle for the bulk of each of the 10K cycle steps.

File: sim/src/main/cc/midasexamples/Driver.cc
Comment (davidbiancolin) [inline sim/src/main/cc/midasexamples/Driver.cc:100]: This isn't required, but vestigial from trying to put the test in midasexamples first?

**Review by davidbiancolin — COMMENTED** [review]
First half

File: sim/src/main/cc/bridges/UARTModule.cc
Comment (davidbiancolin) [inline sim/src/main/cc/bridges/UARTModule.cc:21]: I'm curious, what's your opinion on streams vs printf for output to stderr?

File: sim/src/main/cc/bridges/UARTModule.cc
Comment (davidbiancolin) [inline sim/src/main/cc/bridges/UARTModule.cc:67]: Should future tests in this project also define there own mains like this? I need to add some TracerV features and wanted to write tests rebased against this branch

**Review by davidbiancolin — COMMENTED** [review]
All in all looks good to me. Thanks for taking the initiative here.  A few things pertaining to docs:

Can you skim over the UART bridge walkthrough and make sure it's still pointing at sensible points of the code. Those are hosted here: https://docs.fires.im/en/main/Golden-Gate/Bridge-Walkthrough.html

Can you also add a section for this new target project, following the FASEDtests example here: https://docs.fires.im/en/main/Advanced-Usage/Generating-Different-Targets.html#fased-tests-fasedtests-project

Maybe consider updating this: https://docs.fires.im/en/main/Developer-Docs/GoldenGate-and-Driver-Development.html if necessary 

The rst source for all of this lives under `docs/` in case you weren't aware.

File: sim/src/main/cc/bridges/UARTModule.cc
Comment (nandor) [reply sim/src/main/cc/bridges/UARTModule.cc:67]: For now, let's define the same method everywhere. The entry points will change in subsequent PRs involving the DPI changes and we can introduce a uniform representation there.

File: sim/src/main/cc/bridges/UARTModule.cc
Comment (nandor) [reply sim/src/main/cc/bridges/UARTModule.cc:57]: Thanks for clarifying this!

File: sim/src/main/cc/bridges/UARTModule.cc
Comment (nandor) [reply sim/src/main/cc/bridges/UARTModule.cc:21]: I've used printf since it was used in other parts as well. But overall I would prefer streams.

Moving forward, I would like to move the test checking functions out of `simif_peek_poke_t` into `TestHarness`, adding googletest-like assertions there as well for verification, so there is no need to print anything in tests.

Comment (davidbiancolin) [reply sim/src/main/cc/bridges/UARTModule.cc:23]: Yeah, that sounds ideal. Thanks for the context.

File: sim/src/main/cc/bridges/UARTModule.cc
Comment (davidbiancolin) [inline sim/src/main/cc/bridges/UARTModule.cc:64]: Better, now you should add an iteration bound in case something deadlocks in the target. YOu could instead check that the DUT is making forward progress by calling simif_t::actual_tcycle, but that's a little trickier to tune -- i'll leave it to your discretion. 

This sort of thing can be caught by a host-time max cycle bound in the metasim harness, but you can provide a better error here.

**Review by davidbiancolin — COMMENTED** [review]
One follow up, and then this will be ready to ship.

File: sim/src/test/scala/bridges/BridgeSuite.scala
Comment (davidbiancolin) [inline sim/src/test/scala/bridges/BridgeSuite.scala:38]: I won't let this block the PR, but you should always label return types moving forward.

File: sim/src/test/scala/bridges/BridgeSuite.scala
Comment (davidbiancolin) [inline sim/src/test/scala/bridges/BridgeSuite.scala:61]: Similarly, you want ` : Unit = ` here. 

I'll catch all of this when i get around to cleaning up some these tests.

**Review by davidbiancolin — APPROVED** [review]

**Pull request merged**
